<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#SQL" data-toc-modified-id="SQL-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>SQL</a></span><ul class="toc-item"><li><span><a href="#Знакомство-с-данными" data-toc-modified-id="Знакомство-с-данными-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Знакомство с данными</a></span></li><li><span><a href="#Задание-1" data-toc-modified-id="Задание-1-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Задание 1</a></span></li><li><span><a href="#Задание-2" data-toc-modified-id="Задание-2-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Задание 2</a></span></li><li><span><a href="#Задание-3" data-toc-modified-id="Задание-3-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Задание 3</a></span></li><li><span><a href="#Задание-4" data-toc-modified-id="Задание-4-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Задание 4</a></span></li><li><span><a href="#Задание-5" data-toc-modified-id="Задание-5-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Задание 5</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

### SQL

Цель исследования: проанализировать базу данных, для того чтобы сформулировать ценностное предложение для нового продукта.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

#### Знакомство с данными

In [4]:
query = '''
            SELECT * FROM books LIMIT 1
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [5]:
books

book_id  author_id         title  num_pages publication_date  publisher_id
0        1        546  'Salem's Lot        594       2005-11-01            93

In [6]:
query = '''
            SELECT * FROM authors LIMIT 1
        '''
authors = pd.io.sql.read_sql(query, con = engine)

In [7]:
authors

author_id      author
0          1  A.S. Byatt

In [8]:
query = '''
            SELECT * FROM publishers LIMIT 1
        '''
publishers = pd.io.sql.read_sql(query, con = engine)

In [9]:
publishers

publisher_id publisher
0             1       Ace

In [10]:
query = '''
            SELECT * FROM ratings LIMIT 1
        '''
ratings = pd.io.sql.read_sql(query, con = engine)

In [11]:
ratings

rating_id  book_id    username  rating
0          1        1  ryanfranco       4

In [12]:
query = '''
            SELECT * FROM reviews LIMIT 1
        '''
reviews = pd.io.sql.read_sql(query, con = engine)

In [13]:
reviews

review_id  book_id      username  \
0          1        1  brandtandrea   

                                                text  
0  Mention society tell send professor analysis. ...

#### Задание 1

Посчитайте, сколько книг вышло после 1 января 2000 года.

In [14]:
query = '''
            SELECT COUNT(book_id) FROM books WHERE CAST(publication_date AS date) > '01.01.2000'
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [15]:
books

count
0    819

После 1 января 2000 года вышло 819 книг.

#### Задание 2

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [16]:
query = '''
            SELECT b.book_id,
                   title,
                   count(review_id),
                   AVG(rating)
            FROM books AS b
            LEFT JOIN reviews AS re ON b.book_id = re.book_id
            LEFT JOIN ratings AS ra ON b.book_id = ra.book_id
            GROUP BY b.book_id
            ORDER BY count(review_id)
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [17]:
books

book_id                                              title  count  \
0        387                               Leonardo's Notebooks      0   
1        221                          Essential Tales and Poems      0   
2        672   The Cat in the Hat and Other Dr. Seuss Favorites      0   
3         83    Anne Rice's The Vampire Lestat: A Graphic Novel      0   
4        808                            The Natural Way to Draw      0   
..       ...                                                ...    ...   
995      299  Harry Potter and the Chamber of Secrets (Harry...    480   
996      302  Harry Potter and the Prisoner of Azkaban (Harr...    492   
997      673                             The Catcher in the Rye    516   
998      750                The Hobbit  or There and Back Again    528   
999      948                            Twilight (Twilight  #1)   1120   

          avg  
0    4.000000  
1    4.000000  
2    5.000000  
3    3.666667  
4    3.000000  
..        ...  
995  4.287500  
996  4.414634  
997  3.825581  
998  4.125000  
999  3.662500  

[1000 rows x 4 columns]

Больше всего отзывов на книгу `Twilight` (Сумерки), но оценки за эту книгу не высокие.

Много отзывов на такие книги как Гарри Поттер (Дж. Роулинг) и Хоббит (Дж. Р.Р. Толкиен), которые очень популярны и пользуются большим спросом. 

Также книга Над пропастью во ржи (Д.Д. Сэлинджер) - один из самых продаваемых романов. Ежегодно в мире продается около 250 тысяч экземпляров, а общий объем продаж перевалил за 65 миллионов экземпляров.

#### Задание 3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [18]:
query = '''
            SELECT publisher,
                   COUNT(book_id)
            FROM books AS b
            INNER JOIN publishers AS p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY publisher
            ORDER BY COUNT(book_id) DESC
            LIMIT 1
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [19]:
books

publisher  count
0  Penguin Books     42

Наибольшее число книг выпустило издательство `Penguin Books` (Британское издательство, основанное в 1935 году в Лондоне).

#### Задание 4

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [28]:
query = '''
            SELECT a.author,
                   AVG(ra.rating) as avg_rating
            FROM books AS b
            INNER JOIN ratings AS ra ON b.book_id = ra.book_id
            INNER JOIN authors AS a ON b.author_id = a.author_id
            WHERE b.book_id in (select bo.book_id      
                                from books bo, ratings r 
                                where bo.book_id = r.book_id       
                                group by bo.book_id
                                having count(*) >= 50)
            GROUP BY a.author
            ORDER BY avg_rating DESC
            LIMIT 1
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [29]:
books

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Самая высокая средняя оценка книг у автора `J.K. Rowling/Mary GrandPré` (Джоан Роулинг/Мэри Грандпре).

#### Задание 5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [22]:
query = '''
            WITH u AS (SELECT username
                          FROM ratings
                          GROUP BY username
                          HAVING COUNT(rating) > 50),
            i AS (SELECT COUNT(review_id) AS review,
                         username
                  FROM reviews
                  GROUP BY username)
                          
            SELECT AVG(review)
            FROM i
            INNER JOIN u ON i.username = u.username
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [23]:
books

avg
0  24.333333

Пользователи, которые поставили больше 50 оценок, в среднем написали 24 отзыва.

###  Вывод

Мы изучили базу данных, в которой хранится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

После 1 января 2000 года вышло 819 книг. Больше всего отзывов на книгу Twilight (Сумерки), но оценки за эту книгу не высокие.

Много отзывов на такие книги как Гарри Поттер (Дж. Роулинг) и Хоббит (Дж. Р.Р. Толкиен), которые очень популярны и пользуются большим спросом.

Также книга Над пропастью во ржи (Д.Д. Сэлинджер) - один из самых продаваемых романов. Ежегодно в мире продается около 250 тысяч экземпляров, а общий объем продаж перевалил за 65 миллионов экземпляров.

Наибольшее число книг выпустило издательство Penguin Books (Британское издательство, основанное в 1935 году в Лондоне).

Самая высокая средняя оценка книг у автора J.K. Rowling/Mary GrandPré (Джоан Роулинг/Мэри Грандпре).

Пользователи, которые поставили больше 50 оценок, в среднем написали 24 отзыва.